### PR TITLE
Refactor channels communication to improve concurrency

### PR DESCRIPTION
**Pull request by exekias**
This change mainly splits samba communication from http response
goroutines.

At the "cost" of spending 5 goroutines per virutal connection, we
gain fine gained control on all established connections. Enabling
us to easily do things like timeouts

Also, connection list handling has been spitted into a new file
(connections.go), to ensure correct concurrency handling of map (which
needs synchronization)

### CODE REVIEW

**Comment by exekias** [comment]
I'm not sure why travis is failing here... working on my env, will check, is it happening to you @sixstone-qq ?

**Comment by exekias** [comment]
Ok, found

File: channels/channels.go [hunk 1]
```diff
@@ -157,22 +98,23 @@ func (VirtualConnections VirtualConnections) HandleOut(w http.ResponseWriter, r
...
-				} else {
-					log.Panic(err)
-				}
-			}
+		// PDUs from In Channel
+		case pdu = <-connection.OutChannel.In:
+		// PDUs from Samba Out
+		case pdu = <-connection.SambaChannel.Out:
+		// Read timeout
+		case <-time.After(0.5 * ConnectionTimeout * time.Millisecond):
```
Comment (sixstone-qq) [inline channels/channels.go]: Do you lower the timeout on purpose from 90% to 50%? This will increase the network traffic, do you see any benefit?

File: channels/connections.go [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+// will close the ones it writes to on termination. This way all of them get
+// notified of the EOF in cascade
+type VirtualConnection struct {
+	Cookie             [16]byte
+	AssociationGroupID [16]byte
+	TimeCreated        time.Time
+	Username           string
+
+	// Connected/Finished signals
+	SambaConnected chan bool
```
Comment (sixstone-qq) [inline channels/connections.go]: It would be great to document a little this part or it is unclear the `sync.WaitGroup` thing IMHO. Or making the name more explicit (`FinishedWaitGroup`)? I don't know really if my proposals are correct.

File: channels/channels.go [hunk 3]
```diff
@@ -135,7 +76,7 @@ func (VirtualConnections VirtualConnections) HandleOut(w http.ResponseWriter, r
 	w.(http.Flusher).Flush()
 
 	// Wait for connection established
-	connection.ServerConn = <-connection.Connected
+	<-connection.SambaConnected
```
Comment (sixstone-qq) [inline channels/channels.go]: Adding a timeout here will be done in another PR, right?

File: main.go [hunk 4]
```diff
@@ -1,8 +1,7 @@
 package main
 
 import (
-	"log"
-
+	"github.com/op/go-logging"
```
Comment (sixstone-qq) [inline main.go]: This change may be in another PR, but it is ok to leave as it is.

File: channels/channels.go (hunk 1 above)
Comment (exekias) [inline channels/channels.go]: I just followed the example in the doc, but you are right, from our tests 0.9 should be safe, I'm reverting this

File: channels/channels.go (hunk 3 above)
Comment (exekias) [inline channels/channels.go]: In theory it should not be needed, current approach will fail faster because of channel closing on errors:
https://github.com/zentyal/gorpcproxy/pull/13/files#diff-03663b7fb16baf7927d5d34c6a0deec1R225
https://github.com/zentyal/gorpcproxy/pull/13/files#diff-03663b7fb16baf7927d5d34c6a0deec1R232

File: channels/connections.go (hunk 2 above)
Comment (exekias) [inline channels/connections.go]: It makes sense, I will add some explanation notes here, I was also wondering a way of doing this "cleaner", no success on that yet :)

File: channels/connections.go [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+	connection.SambaChannel.In = make(chan *packets.PDU, 10)
+	connection.SambaChannel.Out = make(chan *packets.PDU, 10)
+	connection.TimeCreated = time.Now()
+
+	// TODO start samba connection already to save some loops?
+	return connection
+}
+
+// GetConnections registered
+func (v *VirtualConnectionsMap) GetConnections() chan *VirtualConnection {
```
Comment (sixstone-qq) [inline channels/connections.go]: According to https://golang.org/doc/effective_go.html#names, the `Get` prefix is not necessary, I would use `Connections`.

File: channels/connections.go [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+		}
+		v.RUnlock()
+		close(res)
+	}()
+
+	return res
+}
+
+// GetNumConnections registered
+func (v *VirtualConnectionsMap) GetNumConnections() int {
```
Comment (sixstone-qq) [inline channels/connections.go]: The same applies here. I'd rather use `ConnectionsNum` or `NumConnections.

File: server/server.go [hunk 7]
```diff
@@ -6,18 +6,23 @@ import (
 	"os"
 
 	// Enable profiler
-	//_ "net/http/pprof"
+	_ "net/http/pprof"
```
Comment (sixstone-qq) [inline server/server.go]: why do you think this change is required to be in master?

Comment (exekias) [inline server/server.go]: I thought we could make this configurable through user settings

Comment (sixstone-qq) [inline server/server.go]: According to https://groups.google.com/forum/#!topic/golang-nuts/fPfbDqnKHdo, conditional import is not possible as the code is already compiled and linked.

Comment (exekias) [inline server/server.go]: Oh true, so my proposal here is to remove pprof before first release
